api/python/provisioner/commands/__init__.py

```diff
@@ -465,20 +474,93 @@ class SetSWUpdateRepo(Set):
     input_type: Type[inputs.SWUpdateRepo] = inputs.SWUpdateRepo
 
-    # TODO rollback
-    def _run(self, params: inputs.SWUpdateRepo, targets: str):
+    @staticmethod
+    def _prepare_repo_for_apply(
+        repo: inputs.SWUpdateRepo, enabled: bool = True
+    ):
         # if local - copy the repo to salt user file root
         # TODO consider to use symlink instead
-        if params.is_local():
-            dest = PRVSNR_USER_FILES_SWUPDATE_REPOS_DIR / params.release
+        if repo.is_local():
+            dest = PRVSNR_USER_FILES_SWUPDATE_REPOS_DIR / repo.release
 
-            if not params.is_dir():  # iso file
+            if not repo.is_dir():  # iso file
                 dest = dest.with_name(dest.name + '.iso')
 
-            copy_to_file_roots(params.source, dest)
+            logger.debug(f"Copying {repo.source} to file roots")
+            copy_to_file_roots(repo.source, dest)
+
+            if not enabled:
+                repo.repo_params = dict(enabled=False)
+
+    def dynamic_validation(self, params: inputs.SWUpdateRepo, targets: str):
+        metadata = {}
+        repo = params
+
+        if repo.is_special():
+            logger.info(
+                "Skipping update repo validation for special value: "
+                f"{repo.source}"
+            )
+            return
+
+        logger.info(
+            f"Validating update repo: release {repo.release}, "
+            f"source {repo.source}"
+        )
+
+        candidate_repo = inputs.SWUpdateRepo(
+            REPO_CANDIDATE_NAME, repo.source
+        )
+
+        # TODO IMPROVE VALIDATION EOS-14350
+        #   - there is no other candidate that is being verified
+
+        # TODO IMPROVE
+        #   - makes sense to try that only on local minion,
+        #     currently it's not convenient since may lead to
+        #     mess in release.sls pillar between all and specific
+        #     minions
+        try:
+            logger.debug(
+                "Configuring update candidate repo for validation"
+            )
+            self._prepare_repo_for_apply(candidate_repo, enabled=False)
+
+            super()._run(candidate_repo, targets)
+
+            # TODO IMPROVE VALIDATION EOS-6078
+            # - yum --disablerepo="*" --enablerepo="sw_update_{candidate_repo.release}" list available  # noqa: E501
+            # - extract metadata from release info file, validate and assign,
+            #   validation:
+            #   - metadata is full enough (at least release information)
+            #   - there is no the same release repo already active EOS-13715
+            #   - (optionally) new version is higher then currently installed
+
+            logger.debug(f"Resolved metadata {metadata}")
+            repo.metadata = metadata
+
+            # TODO IMPROVE EOS-6078
+            # - set 'release' to version from resolved metadata,
+            #   that would make `release` param unnecessary at all
+            #   repo.release = ...
```

Comment:
Why are we not handling the exception here if we are putting it in a try block?

Reply:
we need finally block here mostly (we might consider context manager for that in future).
Exceptions will just go outside. That might be also improved later if needed since it is related to current efforts for messaging improvement.

Reply:
we need finally block here mostly (we might consider context manager for that in future).
Exceptions will just go outside. That might be also improved later if needed since it is related to current efforts for messaging improvement.